### PR TITLE
Add Restart For CountParticles, EnergyFields & EnergyParticles

### PR DESCRIPTION
**Pull request by ax3l**
This avoids loosing the data before the restart, basically the same implementation as #540.

### Minors

- code clean up: "frequency" is a period, `#pragma once`
- refactored common code, also in `BinEnergyParticles`

### To Do

- [x] run time test
- [x] update [wiki](https://github.com/ComputationalRadiationPhysics/picongpu/wiki/PIConGPU-Plugins) (remove the note where implemented)

### CODE REVIEW

File: src/picongpu/include/plugins/CountParticles.hpp [hunk 1]
```diff
@@ -154,6 +155,54 @@ class CountParticles : public ILightweightPlugin
...
+
+        if( outFile.is_open() )
+            outFile.close();
+
+        std::stringstream sStep;
+        sStep << restartStep;
+
+        path src( restartDirectory + std::string("/") + filename +
+        std::string(".") + sStep.str() );
+        path dst( filename );
```
Comment (f-schmitt-zih) [inline src/picongpu/include/plugins/CountParticles.hpp]: are you sure that `.` is a safe directory to copy the file to?

File: src/picongpu/include/plugins/CountParticles.hpp [hunk 2]
```diff
@@ -154,6 +155,54 @@ class CountParticles : public ILightweightPlugin
...
+            std::cerr << "[Plugin] [" << analyzerPrefix
+                      << "] Can't open file '" << filename
+                      << "', output disabled" << std::endl;
+            writeToFile = false;
+        }
+    }
+
+    void checkpoint(uint32_t currentStep, const std::string checkpointDirectory)
+    {
+        if( writeToFile )
```
Comment (f-schmitt-zih) [inline src/picongpu/include/plugins/CountParticles.hpp]: you might want to use the same syntax as in `restart`: `if (!writeToFile)...`

File: src/picongpu/include/plugins/CountParticles.hpp (hunk 1 above)
Comment (ax3l) [inline src/picongpu/include/plugins/CountParticles.hpp]: That is the file name extension, not the dir :)

File: src/picongpu/include/plugins/CountParticles.hpp (hunk 2 above)
Comment (ax3l) [inline src/picongpu/include/plugins/CountParticles.hpp]: Yes a quick return is more sexy, I agree

File: src/picongpu/include/plugins/CountParticles.hpp (hunk 1 above)
Comment (ax3l) [inline src/picongpu/include/plugins/CountParticles.hpp]: Brr, still: wrong spacing!

**Comment by ax3l** [comment]
@psychocoderHPC tested, it works!

**Comment by psychocoderHPC** [comment]
nice pull request!

**Pull request merged**